Share code for handling monitoring events to enable taker

da-kami opened the pull request:
Given the symmetry of the protocol all the state transitions should be exactly the same on the taker and maker side.

State transitions logic moved into the `Cfd` - I propose to do that for all other transitions as well (eventually).

thomaseizinger reviewed (commented):
What do you think of putting these transitions onto the `Cfd` itself? That would allow us to keep the actual transition functions side-effect free.

da-kami committed:
Allow transitions to same state in db

We currently have cases where the CFD might transition to a same state twice when restarting.
Furthermore, we might transition from `Open` to `Open` upon renewal.
Since it is currently unclear if this restriction is really needed we just print a warning in the logs instead of failing.

da-kami committed:
Bundle monitoring events that change cfd state

This goes towards handling all cfd state transitions in one place, so it becomes easier to reason about it.
Currently, state transitions are scattered over the code of the cfd actor.
This change bundles the actual transition in the cfd itself.
Inserting the new state into the db is still in the responsibility of the caller.

note: the `monitor` import in `taker.rs` should technically not be part of this commit, but was needed because I decided that the `CfdMonitoringEvent` should live in the `monitor` actor. But since that is not part of the `Cfd` in `model` the taker needs to know about the `monitor`.

da-kami committed:
move refund transaction signing onto cfd

da-kami committed:
Taker monitors

Wire the monitor into the taker.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -363,6 +364,192 @@ impl Cfd {
+
+        let order_id = self.order.id;
+
+        let new_state = match event {
```
If you wouldn't bundle all the events in an enum but have `handle_lock_finality` etc then you would get rid of one level of indentation within all these functions, potentially making things more readable.

da-kami replied:
The idea was to prepare this so that **all** state transitions would be handle in one function (eventually). 
Granted, there is only one type of events at the moment, but I think it's OK to outline this. 
If the indentation is annoying we can also match on the complete states instead of indenting. Will keep that for now if that's OK.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -363,6 +364,192 @@ impl Cfd {
     #[allow(dead_code)]
     pub const CET_TIMELOCK: u32 = 12;
+
+    pub fn transition_to(&self, event: CfdStateChangeEvent) -> Result<CfdState> {
```
I think there might be a better name for this function. `transition_to` implies that you are passing another state but actually you are passing an `event`. Events are usually "handled".

thomaseizinger commented on daemon/src/monitor.rs:
```diff
@@ -147,33 +142,30 @@ impl xtra::Message for StartMonitoring {
-impl xtra::Message for CetTimelockExpired {
-    type Result = ();
+#[derive(Debug, Clone)]
+pub enum CfdMonitoringEvent {
```
When naming things, try to make use of the modules name you are currently in.

This module is called `monitor`, so referring to this type as `monitor::Event` would give enough context.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -228,12 +232,62 @@ impl Actor {
+                .await?;
+
+            tracing::info!("Refund transaction published on chain: {}", txid);
+        }
```
Didn't we want triggering of these things up to the user?

da-kami replied:
When implementing this with @luckysori we went by the convention: If the user has a choice we bubble it up, if the user does not have a choice we act. Refund is no choice. To my understanding, if the refund timelock expired you have to refund. Or is there a scenario where it is save to do anything but refund then? 

We can still give the user a choice based on the state and only trigger the publication upon user request. Shall I record it as a ticket?

luckysori replied:
The assumption is that you only reach this state if the oracle never attests a price and the refund timelock has expired. At this stage I don't think the application should wait for anything: it should refund automatically.

thomaseizinger reviewed (commented):
Thanks! Definitely an improvement :)

thomaseizinger reviewed (approved):
Meant to approve!

da-kami committed:
Work in review comments

luckysori commented on daemon/src/model/cfd.rs:
```diff
@@ -363,6 +365,221 @@ impl Cfd {
+
+#[derive(Debug, Clone)]
+pub enum CfdStateChangeEvent {
+    // TODO: groupd other events by actors into enums and add them here so we can bundle all
```
:question: Is this TODO up-to-date?

da-kami replied:
It is meant for all other state transitions (outside of monitoring). the idea is to bundle *all* state transitions of a Cfd in this function. 
That said, it is still valid (but has a typo...)

luckysori reviewed (approved):
Thanks for cleaning up, looks great!
